### ISSUE TITLE
Using TARGETARCH before the FROM line does not work

### ISSUE THREAD

**Issue by ndepal** [docker/buildx#574]
I have a Dockerfile where I would like to use the `TARGETARCH` that is set by buildx in my `FROM` line:

```
ARG TARGETARCH

FROM gcr.io/my-container-registry/my-container-${TARGETARCH}

...
```

(I'm aware that if I built `my-container` for multiple platforms in the same `docker buildx build --platform x,y` command, I would not have to use different tags for different platforms. But I don't want to do that since I usually want to build the images for each architecture on machines with that architecture.)

# Expected:
Running `docker buildx build --platform arm64 ...` should result in `FROM gcr.io/my-container-registry/my-container-arm64`

# What actually happens:
`failed to parse stage name "gcr.io/my-container-registry/my-container-": invalid reference format`

Using `ARG TARGETARCH` after the `FROM` line works as expected (although I did also run into https://github.com/docker/buildx/issues/510, expecting it to work as @MarcelWaldvogel did).

Is there a way to get this to work? Or is this the intended behavior?

**Comment by MarcelWaldvogel** [comment]
The [documentation](https://docs.docker.com/buildx/working-with-buildx/#build-multi-platform-images) says (emphasis mine):

> A list of build arguments like `BUILDPLATFORM` and `TARGETPLATFORM` is available automatically inside your Dockerfile and can be leveraged **by the processes running as part of your build.**

So, it seems to be the documented behavior. Even though I completely understand that it is not what is expected.

I guess you have to run a preprocessor over the Dockerfile to create a per-platform Dockerfile. (And then merge things in the end, if merging into a single multi-platform file is required.) (Or creating a virtual multi-platform image of your base image using [`docker manifest`](https://www.docker.com/blog/multi-arch-build-and-images-the-simple-way/).)

**Comment by tonistiigi** [comment]
These args are available automatically in `FROM` scope, you don't need to redefine them.

Just:

```
FROM gcr.io/my-container-registry/my-container-${TARGETARCH}
```
will work fine.

But I agree that it is weird that redefinition without a default value would cause the behavior to change and should be fixed. Care to move this to buildkit repo for tracking after validating that it works apart from redefinition overriding the automatic value.

**Comment by ndepal** [comment]
> The documentation says (emphasis mine):

Fair point, I guess I did not read "part of your build" to mean only my RUN etc instructions. 

> These args are available automatically in FROM scope, you don't need to redefine them.

That is indeed true, very cool! Thanks for pointing that out @tonistiigi. This behavior is confusing to me though, due to this sentence in the [documentation](https://docs.docker.com/engine/reference/builder/#automatic-platform-args-in-the-global-scope):
> These arguments are defined in the global scope so are not automatically available inside build stages or for your RUN commands. To expose one of these arguments inside the build stage redefine it without value.

It appears that the `FROM` line is treated fundamentally differently from other lines in Dockerfiles, that distinction is not clear to me. Is this documented somewhere?

@MarcelWaldvogel That's a nice article, thanks! I have `buildx` set up and working in principle.

The issue I have with this approach is that building the arm64 image on my amd64 machine takes ages, so I'd much rather build the images on their respective platforms for faster builds.

But then I end up with images in my registry that are each only good for one architecture. So I would have to manually create the manifest list with e.g.
```
docker manifest create \
    gcr.io/my-container-registry/my-container:latest \
    gcr.io/my-container-registry/my-container-arm64:latest \
    gcr.io/my-container-registry/my-container-amd64:latest
docker manifest push gcr.io/my-container-registry/my-container:latest
```
after both `docker buildx` commands have finished. And I could only do this after I build them for both architectures. So I can't rebuild an image only on amd64 and then re-run my `docker-compose` to get the new tag, because that tag does not exist until I can create the merged manifest.

What I also don't like about that is that I'm just claiming some two images have the same content (except for the architecture), but that might not be true, so it's a source of error.

So I don't really see what I'm gaining from having all these images in the same registry. If I put the arch into the image name, I will always be sure that that I'm pulling the tag and architecture I intend to.

The only downside to this is that in my docker-compose, I now have a `TARGETARCH` env variable that I append to all the `image: ...` specifications. But this seems like less work than having to "combine" pushed images with `docker manifest create` whenever they were changed.

Or am I missing something and this is all a lot simpler than I think?

**Comment by tonistiigi** [comment]
>  that distinction is not clear to me. Is this documented somewhere?

https://docs.docker.com/engine/reference/builder/#understand-how-arg-and-from-interact

> The issue I have with this approach is that building the arm64 image on my amd64 machine takes ages

You can also add multiple nodes to a single buildx builder instance with `--append` and build on multiple nodes together without any manual merging.

**Comment by MarcelWaldvogel** [comment]
@ndepal I use a local docker build (using only the local architecture and no publishing to a registry) for all the development+testing. The multi-architecture build only happens on releases, so the emulation speed is not an issue. Your processes might differ; but then you have the choice of
* dedicated builder nodes,
* `docker manifest`
* `docker buildx --append`.
I hope one of them suits your needs.

**Comment by ndepal** [comment]
> docs.docker.com/engine/reference/builder/#understand-how-arg-and-from-interact

Thanks for the link, that is a helpful page. Although, FWIW, if you had not told me that I must not redefine the ARG before the FROM line, I would not have pieced it together on that page.

The `--append` option looks very interesting, I'll try it out!

Thanks to both of you for your help.
